packages/nextjs-mf/src/plugins/NextFederationPlugin.ts

```diff
@@ -38,187 +52,470 @@ const regexEqual = (x, y) => {
+          /pages[\\/]middleware/,
+          /pages[\\/]api/,
+        ],
+        resourceQuery: (query: string) => !query.includes('hasBoundary'),
+        use: [
+          //@ts-ignore
+          ...loaderChain,
+          //@ts-ignore
+          {
+            loader: path.resolve(__dirname, '../loaders/async-boundary-loader'),
+          },
+        ],
+      });
     }
+  }
+}
 
-    if (!['server', 'client'].includes(compiler.options.name)) {
-      return;
+/**
+ * Apply remote delegates.
+ */
+function applyRemoteDelegates(options, compiler) {
+  if (options.remotes) {
+    const delegates = getDelegates(options.remotes);
+
+    compiler.options.module.rules.push({
+      enforce: 'pre',
+      test: [/internal-delegate-hoist/, /delegate-hoist-container/],
+      include: [
+        compiler.context,
+        /internal-delegate-hoist/,
+        /delegate-hoist-container/,
+        /next[\\/]dist/,
+      ],
+      loader: path.resolve(__dirname, '../loaders/delegateLoader'),
+      options: {
+        delegates,
+      },
+    });
+
+    if (delegates && Object.keys(delegates).length > 0) {
+      const knownDelegates = Object.entries(delegates).map(([name, remote]) => {
+        const delegate = remote.replace('internal ', '').split('?')[0];
+        return delegate;
+      });
+
+      if (this._options.exposes) {
+        compiler.options.module.rules.push({
+          enforce: 'pre',
+          test(request: string) {
+            const found = knownDelegates.some((delegate) => {
+              return request.includes(delegate);
+            });
+
+            return found;
+          },
+          loader: path.resolve(__dirname, '../loaders/inject-single-host'),
+          options: {
+            name: this._options.name,
+          },
+        });
+      }
     }
+  }
+}
+
+// Create runtime variables.
+//@ts-ignore
+function createRuntimeVariables(webpack) {
+  new webpack.DefinePlugin({
+    'process.env.REMOTES': createRuntimeVariables(this._options.remotes),
+    'process.env.CURRENT_HOST': JSON.stringify(this._options.name),
+  }).apply(compiler);
+}
+
+/**
+ * NextFederationPlugin is a webpack plugin that handles Next.js application
+ * federation using Module Federation.
+ */
+export class NextFederationPlugin {
+  _options: ModuleFederationPluginOptions;
+  _extraOptions: NextFederationPluginExtraOptions;
+
+  /**
+   * Constructs the NextFederationPlugin with the provided options.
+   *
+   * @param options The options to configure the plugin.
+   */
+  constructor(options: NextFederationPluginOptions) {
+    const { mainOptions, extraOptions } = setOptions(options);
+    this._options = mainOptions;
+    this._extraOptions = extraOptions;
+  }
+
+  apply(compiler: Compiler) {
+    // Validate the compiler options
+    const validCompile = validateCompilerOptions(compiler);
+    if (!validCompile) return;
+    // Validate the NextFederationPlugin options
+    validatePluginOptions(this._options);
```

Comment:
shouldn't we pass `this._extraOptions` since we want to validate `NextFederationPlugin` options?

Reply:
it only validates the module federation parts that webpack needs and ecosystem needs. extra options have little impact on core functionality